File: examples/lookup.dart
import 'dart:io';
import 'dart:typed_data';

class Feed {
  final String name;
  final int baseScore, confidence;
  final int flagsMask, categoriesMask;
  final List<int> ipv4Starts, ipv4Ends;
  final List<String> flags, categories;
  Feed(this.name, this.baseScore, this.confidence, this.flagsMask,
      this.categoriesMask, this.ipv4Starts, this.ipv4Ends,
      this.flags, this.categories);
}

class Reader {
  final ByteData data;
  int pos;
  Reader(this.data, [this.pos = 0]);

  int u8() => data.getUint8(pos++);
  int u16() { var v = data.getUint16(pos, Endian.little); pos += 2; return v; }
  int u32() { var v = data.getUint32(pos, Endian.little); pos += 4; return v; }

  String str() {
    var len = u8();
    var s = String.fromCharCodes(
        data.buffer.asUint8List(data.offsetInBytes + pos, len));
    pos += len;
    return s;
  }

  int varint() {
    int result = 0, shift = 0;
    while (true) {
      var b = u8();
      result |= (b & 0x7F) << shift;
      if (b & 0x80 == 0) return result;
      shift += 7;
    }
  }
}

List<Feed> load([String path = 'blocklist.bin']) {
  var bytes = File(path).readAsBytesSync();
  var r = Reader(ByteData.sublistView(bytes), 4);
  r.u8(); // version
  r.u32(); // timestamp

  var fc = r.u8();
  var flags = List.generate(fc, (_) => r.str());
  var cc = r.u8();
  var cats = List.generate(cc, (_) => r.str());

  var feedCount = r.u16();
  var feeds = <Feed>[];

  for (var i = 0; i < feedCount; i++) {
    var name = r.str();
    var bs = r.u8(), co = r.u8();
    var fm = r.u32(), cm = r.u8();
    var rc = r.u32();

    var v4s = <int>[], v4e = <int>[];
    var current = 0;
    for (var j = 0; j < rc; j++) {
      current += r.varint();
      var end = current + r.varint();
      if (end <= 0xFFFFFFFF) { v4s.add(current); v4e.add(end); }
    }

    var mf = [for (var k = 0; k < flags.length; k++)
      if (fm & (1 << k) != 0) flags[k]];
    var mc = [for (var k = 0; k < cats.length; k++)
      if (cm & (1 << k) != 0) cats[k]];

    feeds.add(Feed(name, bs, co, fm, cm, v4s, v4e, mf, mc));
  }
  return feeds;
}

int bisectRight(List<int> arr, int target) {
  var lo = 0, hi = arr.length;
  while (lo < hi) {
    var mid = (lo + hi) >> 1;
    if (arr[mid] <= target) lo = mid + 1; else hi = mid;
  }
  return lo;
}

int? parseIPv4(String s) {
  var parts = s.split('.');
  if (parts.length != 4) return null;
  try {
    return (int.parse(parts[0]) << 24) | (int.parse(parts[1]) << 16)
         | (int.parse(parts[2]) << 8) | int.parse(parts[3]);
  } catch (_) { return null; }
}

void main(List<String> args) {
  if (args.isEmpty) {
    stderr.writeln('Usage: dart lookup.dart <ip> [<ip> ...]');
    return;
  }

  var feeds = load();

  for (var ip in args) {
    var target = parseIPv4(ip);
    if (target == null) { print('$ip: invalid IP'); continue; }

    var found = false;
    for (var f in feeds) {
      var idx = bisectRight(f.ipv4Starts, target) - 1;
      if (idx >= 0 && target <= f.ipv4Ends[idx]) {
        var score = (f.baseScore / 200.0) * (f.confidence / 200.0);
        var parts = [f.name, 'score=${score.toStringAsFixed(2)}'];
        if (f.flags.isNotEmpty) parts.add('flags=${f.flags.join(",")}');
        if (f.categories.isNotEmpty)
          parts.add('cats=${f.categories.join(",")}');
        print('$ip: ${parts.join(" | ")}');
        found = true;
      }
    }
    if (!found) print('$ip: no matches');
  }
}

File: examples/lookup.nim
import sys, os, strutils, algorithm

type Feed = object
  name: string
  baseScore, confidence: uint8
  flagsMask: uint32
  categoriesMask: uint8
  ipv4Starts, ipv4Ends: seq[uint64]
  flags, categories: seq[string]

proc readVarint(data: openArray[uint8], pos: var int): uint64 =
  var shift = 0
  while true:
    let b = data[pos]; inc pos
    result = result or (uint64(b and 0x7F) shl shift)
    if (b and 0x80) == 0: return
    shift += 7

proc readStr(data: openArray[uint8], pos: var int): string =
  let length = int(data[pos]); inc pos
  result = newString(length)
  for i in 0..<length: result[i] = char(data[pos + i])
  pos += length

proc readU16(data: openArray[uint8], pos: var int): uint16 =
  result = uint16(data[pos]) or (uint16(data[pos+1]) shl 8)
  pos += 2

proc readU32(data: openArray[uint8], pos: var int): uint32 =
  result = uint32(data[pos]) or (uint32(data[pos+1]) shl 8) or
           (uint32(data[pos+2]) shl 16) or (uint32(data[pos+3]) shl 24)
  pos += 4

proc load(path = "blocklist.bin"): seq[Feed] =
  let data = cast[seq[uint8]](readFile(path))
  var pos = 4
  assert data[pos] == 2; inc pos
  pos += 4 # timestamp

  let fc = int(data[pos]); inc pos
  var flagTable = newSeq[string](fc)
  for i in 0..<fc: flagTable[i] = readStr(data, pos)

  let cc = int(data[pos]); inc pos
  var catTable = newSeq[string](cc)
  for i in 0..<cc: catTable[i] = readStr(data, pos)

  let feedCount = int(readU16(data, pos))
  result = newSeq[Feed](feedCount)

  for fi in 0..<feedCount:
    var f: Feed
    f.name = readStr(data, pos)
    f.baseScore = data[pos]; inc pos
    f.confidence = data[pos]; inc pos
    f.flagsMask = readU32(data, pos)
    f.categoriesMask = data[pos]; inc pos
    let rc = int(readU32(data, pos))

    var current: uint64 = 0
    for _ in 0..<rc:
      current += readVarint(data, pos)
      let size = readVarint(data, pos)
      let en = current + size
      if en <= 0xFFFFFFFF'u64:
        f.ipv4Starts.add current
        f.ipv4Ends.add en

    for i in 0..<flagTable.len:
      if (f.flagsMask and (1'u32 shl i)) != 0:
        f.flags.add flagTable[i]
    for i in 0..<catTable.len:
      if (f.categoriesMask and (1'u8 shl i)) != 0:
        f.categories.add catTable[i]

    result[fi] = f

proc bisectRight(arr: seq[uint64], target: uint64): int =
  var lo = 0; var hi = arr.len
  while lo < hi:
    let mid = lo + (hi - lo) div 2
    if arr[mid] <= target: lo = mid + 1
    else: hi = mid
  lo

proc parseIPv4(s: string): uint64 =
  let parts = s.split('.')
  assert parts.len == 4
  for p in parts:
    result = (result shl 8) or uint64(parseInt(p))

when isMainModule:
  if paramCount() < 1:
    quit "Usage: lookup <ip> [<ip> ...]"

  let feeds = load()

  for i in 1..paramCount():
    let ip = paramStr(i)
    var target: uint64
    try: target = parseIPv4(ip)
    except: echo ip, ": invalid IP"; continue

    var found = false
    for f in feeds:
      let idx = bisectRight(f.ipv4Starts, target) - 1
      if idx >= 0 and target <= f.ipv4Ends[idx]:
        let score = (float(f.baseScore) / 200.0) *
                    (float(f.confidence) / 200.0)
        var parts = @[f.name, "score=" & formatFloat(score, ffDecimal, 2)]
        if f.flags.len > 0:
          parts.add "flags=" & f.flags.join(",")
        if f.categories.len > 0:
          parts.add "cats=" & f.categories.join(",")
        echo ip, ": ", parts.join(" | ")
        found = true

    if not found:
      echo ip, ": no matches"
